### PR TITLE
Make node_id required in GetNodeIDResponse

### PR DESCRIPTION
**Pull request by saad-ali**
* Make `node_id` a required field in the `GetNodeIDResponse` if the plugin has `PUBLISH_UNPUBLISH_VOLUME` controller capability.
* Make `node_id` a required field in the `ControllerPublishVolume` and `ControllerUnpublishVolume` calls.
* Require CO to pass it's `node_id` in `GetNodeIdReqeust` so that SP's that don't have their own custom internal naming scheme do not have to implement complicated logic to figure out the node name. 

Fixes https://github.com/container-storage-interface/spec/issues/121

### CODE REVIEW

File: csi.proto [hunk 1]
```diff
@@ -332,15 +330,8 @@ message ControllerUnpublishVolumeRequest {
...
-  // clear) this field to match the node ID returned by `GetNodeID`.
-  // `GetNodeID` is allowed to omit node ID from a successful response;
-  // in such cases the CO SHALL NOT specify this field.
-  //
-  // If `GetNodeID` does not omit node ID from a successful response,
-  // the CO MAY omit this field as well, indicating that it does not
-  // know which node the volume was previously used. The Plugin SHOULD
-  // return an Error if this is not supported.
+  // The ID of the node. This field is REQUIRED. The CO SHALL set this
+  // field to match the node ID returned by `GetNodeID`.
```
Comment (jdef) [inline csi.proto]: We killed a use case here by making this REQUIRED: a CO may intentionally leave this field blank to indicate to the plugin that the volume should be unpublished from every node in the cluster. Is killing that use case really our intent here?

File: csi.proto [hunk 2]
```diff
@@ -576,13 +567,18 @@ message NodeUnpublishVolumeResponse {}
 message GetNodeIDRequest {
   // The API version assumed by the CO. This is a REQUIRED field.
   Version version = 1;
+
+  // The ID of the node as understood by the CO which MAY be returned by
+  // the SP in the GetNodeIDResponse if the SP does not implement a
+  // custom node naming scheme and understands (or internally maps) the
+  // CO's node id.
```
Comment (jdef) [inline csi.proto]: REQUIRED? OPTIONAL?

File: csi.proto (hunk 1 above)
Comment (jieyu) [reply csi.proto]: Yeah. I think we deliberately made this field optional to capture the case where the CO just want to detach the volume from any node.  I still believe this is useful in some reconciliation scenarios.

Comment (saad-ali) [reply csi.proto]: Ack. I'll restore this then.

File: csi.proto (hunk 2 above)
Comment (saad-ali) [reply csi.proto]: Will add.

**Comment by saad-ali** [comment]
@jdef @jieyu feedback addressed, PTAL

File: spec.md [hunk 3]
```diff
@@ -1132,13 +1126,18 @@ The result of this call will be used by CO in `ControllerPublishVolume`.
 message GetNodeIDRequest {
   // The API version assumed by the CO. This is a REQUIRED field.
   Version version = 1;
+
+  // The ID of the node as understood by the CO which MAY be returned by
```
Comment (jieyu) [inline spec.md]: cc @akutz 

This is still controversial. Can we make this change a separate PR?

Comment (saad-ali) [reply spec.md]: Done. Will pull this in to a separate PR.

**Comment by saad-ali** [comment]
PTAL

File: spec.md [hunk 4]
```diff
@@ -1135,10 +1129,9 @@ message GetNodeIDRequest {
 }
 
 message GetNodeIDResponse {
-  // The ID of the node which SHALL be used by CO in
-  // `ControllerPublishVolume`. This is an OPTIONAL field. If unset,
-  // the CO SHALL leave the `node_id` field unset in
-  // `ControllerPublishVolume`.
+  // The ID of the node as understood by the SP which SHALL be used by
+  // CO in subsequent `ControllerPublishVolume`.
+  // This is a REQUIRED field.
```
Comment (jdef) [inline spec.md]: suggestion: This is an OPTIONAL field that is REQUIRED only if the controller advertises the `PUBLISH_UNPUBLISH_VOLUME` capability.

because not implementing the call means that GetNodeIDResponse is empty, and that violates the REQUIRED semantics as currently documented.

Comment (saad-ali) [reply spec.md]: https://github.com/container-storage-interface/spec/pull/152
